### PR TITLE
Ennemy waves

### PR DESCRIPTION
**Pull request by mangasteak**
## Description
Adding ennemy wave to dev

## Lien tache Jira
https://leomaman.atlassian.net/browse/TYPE-122?atlOrigin=eyJpIjoiZjIwOThkY2ZjM2I5NDc1ZjgwZWExNmE0ODY5YTQxOTQiLCJwIjoiaiJ9

## Changement
instead of one enemy there now multiple wave of enemies spawning every 5 seconds

## Summary by CodeRabbit

New Features:
- Introduced a new `WaveSystem` class to manage and generate waves of enemy entities in the game, enhancing gameplay dynamics.
- Added `WaveSystem` to the server core systems, integrating the new feature into the game server.

Refactor:
- Removed the creation and registration of `EnemyEntity` from `ServerCore.cpp`, streamlining entity management.

Style:
- Added a new namespace `ECS` to `EnemyEntity.hpp`, improving code organization.

Chores:
- Updated `CMakeLists.txt` to include new files "WaveSystem.cpp" and "WaveSystem.hpp".
- Modified exit codes and error messages in `install.sh` for better error handling.

### CODE REVIEW

File: server/ServerCore.cpp [hunk 1]
```diff
@@ -28,18 +29,18 @@ ECS::ServerCore::ServerCore(RType::Server &server) : _server(server)
...
     _systems.push_back(std::make_unique<ECS::CollisionSystem>());
     _systems.push_back(std::make_unique<ECS::GameSystem>());
+    _systems.push_back(std::make_unique<ECS::WaveSystem>(_entityFactory));
 }
 
 void ECS::ServerCore::_initEntities()
 {
     std::shared_ptr<ECS::Entity> player = std::make_shared<PlayerEntity>();
-    std::shared_ptr<ECS::Entity> enemy = std::make_shared<EnemyEntity>(0);
+    //std::shared_ptr<ECS::Entity> enemy = std::make_shared<EnemyEntity>(0);
```
Comment (christophechr) [inline server/ServerCore.cpp:38]: plz remove all comments

File: server/ServerCore.cpp [hunk 2]
```diff
@@ -28,18 +29,18 @@ ECS::ServerCore::ServerCore(RType::Server &server) : _server(server)
...
     std::shared_ptr<ECS::Entity> playerBullet = std::make_shared<PlayerBullet>(0);
 
-    std::cout << "At creation : x: " << enemy->getComponent<PositionComponent>()->x << "; y: " << enemy->getComponent<PositionComponent>()->y << std::endl;
-
+    //std::cout << "At creation : x: " << enemy->getComponent<PositionComponent>()->x << "; y: " << enemy->getComponent<PositionComponent>()->y << std::endl;
```
Comment (christophechr) [inline server/ServerCore.cpp:41]: same here

File: server/ServerCore.cpp [hunk 3]
```diff
@@ -28,18 +29,18 @@ ECS::ServerCore::ServerCore(RType::Server &server) : _server(server)
...
     _entityFactory.registerEntity(player, "player");
-    _entityFactory.registerEntity(enemy, "entity" + std::to_string(ECS::Entity::ENEMY_CLASSIC));
+    //_entityFactory.registerEntity(enemy, "entity" + std::to_string(ECS::Entity::ENEMY_CLASSIC));
```
Comment (christophechr) [inline server/ServerCore.cpp:43]: and here

**Review by christophechr — CHANGES_REQUESTED** [review]

**Review by christophechr — APPROVED** [review]

File: build.sh
Comment (raphaelMrci) [inline build.sh:1]: Pk il est la so ?

File: server/ecs/WaveSystem.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        packetQueue.push_back(*packet);
+        sceneManager.getCurrentScene()->addEntity(entity);
+    }
+}
+
+std::vector<std::shared_ptr<ECS::Entity>> ECS::WaveSystem::getWave(int i)
+{
+    std::vector<std::shared_ptr<ECS::Entity>> result;
+
+    for (int j = 0; j < _waves[i]; j++) {
```
Comment (raphaelMrci) [inline server/ecs/WaveSystem.cpp:114]: pas de check pour _waves[i], peut etre en faire un ?

Comment (raphaelMrci) [reply server/ecs/WaveSystem.cpp:114]: not mandatory

File: server/ecs/WaveSystem.hpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        public:
+            WaveSystem(const EntityFactory &Factory) ;
+
+            void update(SceneManager &sceneManager, float deltaTime, std::vector<Network::Packet> &packetQueue) override;
+            std::vector<std::shared_ptr<ECS::Entity>> getWave(int i);
+            void createEnemy(int WaveCount, int EnemyCount, int x, int y);
+
+        protected:
+        private:
+            EntityFactory _factory;
```
Comment (raphaelMrci) [inline server/ecs/WaveSystem.hpp:29]: Ca doit etre une reference, pas une copie, il FAUT mettre une esperluette

File: server/ecs/WaveSystem.hpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+            std::vector<int> _waves;
+            float timer;
+    };
+}
+
+
+#endif /* !WAVESYSTEM_HPP_ */
```
[the remark this answers is not in the document]
Comment (raphaelMrci) [reply server/ecs/WaveSystem.hpp:36]: @mangasteak GG

File: build.sh
Comment (Codrux2200) [inline build.sh:1]: a supprimer

File: install.sh
Comment (Codrux2200) [inline install.sh:1]: c'est pas la bonne version sa va tout casser

**Review by Codrux2200 — APPROVED** [review]

**Pull request merged**